source-firestore/pull.go

```diff
@@ -118,6 +132,15 @@ type resourceState struct {
 	// to be performed to re-establish consistency.
 	Inconsistent bool `json:"Inconsistent,omitempty"`
 
+	// The `RestartCursorValue` property is updated periodically during change streaming
+	// (if the binding has a `RestartCursorPath` setting) so it holds an arbitrarily selected
+	// restart-cursor value which was recently observed in the dataset.
+	RestartCursorValue any `json:"RestartCursorValue,omitempty"`
+
+	restartCursorPath           []string  // The parsed path to the restart-cursor property. Not serialized as part of the state, it's just copied from the resource spec for convenience.
+	pendingRestartCursorTimeout time.Time // The timestamp after which the pending restart-cursor value may be promoted to active. Zero if there is no pending value.
```

Comment:
I'm likely missing something here, but why do we need to wait until some amount of time has passed before promoting a restart cursor?

Reply:
The point of the whole "restart cursor" thing isn't precision so much as it is cutting the bulk of the unnecessarily redundant backfill work we would otherwise have to do when the stream breaks. And I can hardly overstate how little I trust Firestore to behave in a remotely sane manner, so we don't want to assume that the changes we observe via streaming will occur in a reliable order.

The failure mode which could potentially occur without the whole "pending value" thing would look like this:

- After a task restart we begin streaming changes, and fail to catch up to that initial consistent point so the stream breaks, and then we have to go re-backfill it.
- But it's entirely possible that Firestore could have sent us those changes in some arbitrary order which doesn't match the order the changes were made in.
  - This would be valid from their standpoint because the way the change streaming protocol is designed we're clearly intended to buffer the entire "initial snapshot" set of changes and get to a consistent point before doing anything with them
  - That's what the official client library does for example, and avoiding that is why we have to drop down and use the raw gRPC protocol directly.
- So if we simply saved the most recent value we observed and then began backfilling from that value after the stream broke, we could potentially skip over some number of changes with slightly smaller values which would have been sent to us later on if the stream hadn't broken.

That would be pretty bad since the entire point of re-backfilling the collection is to make sure it's consistent with the source dataset, so we're deliberately biasing things so that we accept a small amount of overlap between the documents observed by change streaming and the documents we'll fetch when re-backfilling, in order to minimize the chance that we miss anything.

Reply:
Makese sense; thanks!